### DISCUSSION TITLE
Webp sub-region loading produces garbled image

### DISCUSSION THREAD

**Discussion by Julianiolo** [libvips/libvips#4041] [General]
**Describe the bug**
When loading a webp image and preparing a region that is smaller than the whole image, the resulting image data is garbled.

When adding an alpha channel via `vips_addalpha`, it fixes the issue (see code.)

Peculiarly, when loading an webp image with an alpha channel ([example](https://www.gstatic.com/webp/gallery3/1_webp_ll.webp)), it still produces the bug. When then doing the `vips_addalpha` it also fixes it! (although the image then has 5 bands, no errors when saving to png though).

This is definitely a issue when loading the image, not when saving it, as I had the same issue appear when processing the region contents with different librarys.

Note that when loading a region that is exactly the size of the image, the bug does not appear. I've also tested using the `access=VIPS_ACCESS_SEQUENTIAL` flag, which does not seem to change anything.

**To Reproduce**
1. Use Image https://www.gstatic.com/webp/gallery/1.sm.webp (Produced bug with every webp image I tested)
2. Use code:
```c
int main(int argc, char** argv) {
    if(VIPS_INIT(argv[0]))
        vips_error_exit("unable to start vips");

    VipsImage* img = vips_image_new_from_file("file.webp", NULL);
    assert(img);
    printf("%dx%d, b:%d, f:%d, c:%d\n", vips_image_get_width(img), vips_image_get_height(img), vips_image_get_bands(img), vips_image_get_format(img), vips_image_get_coding(img));

    // uncomment to fix
    // VipsImage* img2;
    // assert(vips_addalpha(img, &img2, NULL) == 0);
    // img = img2;

    VipsRegion* region = vips_region_new(img);
    assert(region);
    VipsRect rect;
    rect.left = 0;
    rect.top = 0;
    rect.width = 128;
    rect.height = 128;

    assert(vips_region_prepare(region, &rect) == 0);

    VipsImage* out = vips_image_new_from_memory(
        VIPS_REGION_ADDR_TOPLEFT(region), VIPS_REGION_SIZEOF_PEL(region)*128*128, 
        128, 128, vips_image_get_bands(img), vips_image_get_format(img)
    );
    assert(out);
    assert(vips_image_write_to_file(out, "out.png", NULL) == 0);

    printf("done!\n");
}
```
3. See below

| Whole Image  | Expected | Actual | info |
| -------- | ------- | -  | - |
| ![img](https://www.gstatic.com/webp/gallery/1.webp) | ![out](https://github.com/user-attachments/assets/3a5ec0cf-f639-4a02-b9bd-02347f5fda31) | ![out](https://github.com/user-attachments/assets/86536285-6f23-4572-b86b-f6ba2651b386) | `320x214, b:3, f:0, c:0` |
| ![img](https://www.gstatic.com/webp/gallery3/1_webp_ll.webp) | ![out](https://github.com/user-attachments/assets/5073dcf1-cbae-4a7a-a1ff-9db343d55dbc) | ![out](https://github.com/user-attachments/assets/83763389-7050-4726-acd4-539b4b407a0f) | `400x301, b:4, f:0, c:0` |

(Expected was obtained via the `vips_addalpha` fix)

**Environment**
 - OS: Windows 10
 - Compiler: msvc 193
 - Vips: 8.15.2

**Comment by Julianiolo** [comment]
Additionally:


When loading the whole image in these smaller regions (so as tiles basically), it looks like this:

(this is the first test image, again with 128x128 tiles)

![image](https://github.com/user-attachments/assets/a847ea53-7198-4eef-9148-4f2fda7164d3)

With gaps for seeing the tiles:
![image](https://github.com/user-attachments/assets/02e355ca-f48b-48c0-9c5a-829f88187e64)

With 32x32 tiles:

![image](https://github.com/user-attachments/assets/9acd950d-fb35-4653-be8e-59f2aa2ee209)

So it seems like the tiles are somewhat repeating horizontally? But not completely, see the top row of the 32x32 tiles.

**Comment by jcupitt** [comment]
Hi @Julianiolo, 

This is expected behaviour --- regions aren't guaranteed to be contiguous in memory. 

When you do `vips_image_new_from_file()` on a webp, it'll decode the whole image into RAM (if the decoded image is under 100mb), since libwebp only supports one shot decode.

Now when you do `vips_region_prepare()`, it'll set the region up as a view of this memory area, it won't copy any pixels. You need to use the `lskip` (line skip, usually called line stride I think in other image libraries) member to go down a scanline.

If you want to do `new_from_memory()` on a region, you'll need to check `lskip` and if it's different from the line length (ie. non-contiguous), you'll need to allocate a memory area and copy lines over. 

A simple way to go this is to make a region on the image, call `vips_region_buffer()` to make it into a memory region, then use `vips_region_copy()` to copy over the pixels.

Alternatively, you can avoid the region machinery and just do `vips_crop()` and `vips_pngsave()` (this is what `dzsave` does), though that can be a bit slow for very small tiles.

**Comment by Julianiolo** [comment]
@jcupitt oh ok, I see, thanks for the help :)
